canonicalwebteam/blog/wordpress_api.py

```diff
@@ -19,71 +21,78 @@ def process_response(response):
     return response.json()
 
 
-def build_get_articles_url(
-    tags=[],
-    per_page=12,
-    page=1,
-    tags_exclude=[],
-    exclude=[],
-    categories=[],
-    sticky="",
-    groups=[],
-    after="",
-    before="",
-    author="",
-):
+def build_url(endpoint, params={}):
     """
     Build url to fetch articles from Wordpress api
-    :param tags: Array of tag ids to fetch articles for
-    :param per_page: Articles to get per page
-    :param page: Page number to get
-    :param tags_exclude: Array of IDs of tags that will be excluded
-    :param exclude: Array of article IDs to be excluded
-    :param category: Array of categories, which articles
-        should be fetched for
-    :param sticky: string 'true' or 'false' to only get featured articles
-    :param before: ISO8601 compliant date string to limit by date
-    :param after: ISO8601 compliant date string to limit by date
+    :param endpoint: The REST endpoint to fetch data from
+    :param params: Dictionary of parameter keys and their values
 
     :returns: URL to Wordpress api
     """
-    url = (
-        f"{API_URL}/posts?per_page={per_page}"
-        f"&tags={','.join(str(id) for id in tags)}"
-        f"&page={page}"
-        f"&group={','.join(str(id) for id in groups)}"
-        f"&tags_exclude={','.join(str(id) for id in tags_exclude)}"
-        f"&categories={','.join(str(id) for id in categories)}"
-        f"&exclude={','.join(str(id) for id in exclude)}"
-        f"&author={author}&_embed"
+
+    def list_to_comma_separated_string(key, value):
+        if type(value) is list:
+            return (key, (",".join(str(item) for item in value)))
+        else:
+            return (key, value)
+
+    params = {key: value for key, value in params.items() if value}
+    params = dict(
+        map(lambda kv: list_to_comma_separated_string(*kv), params.items())
```

Comment:
I still feel you're over-engineering this. Why isn't a simple `for key, value in params.items()` sufficient? Lambdas, inline functions, lists of tuples and dictionary comprehensions are just fundamentally confusing constructs for people to read when a much more simple construct would work just as well.

E.g., in this case how is what you've written any different from what I previously suggested?

``` python3
for key, value in params.items()
    if type(value) is list:
        params[key] = ",".join(str(item) for item in value)
    elif not value:
        del params[key]
```

I also would rather that you were more explicit about what sorts of items are going to get removed from the dictionary (rather than simply `not value`), as I personally find that quite a confusing behaviour.

For exampe, you mentioned empty strings. I would expect that if I did something like `build_query({"a": "1", "b": False, "c": ""})` then what I'd get would be `"a=1&b=false&c`. I understand that for your specific use-case in `get_articles` you want to be able to pass through `None` for some of the values and have them not end up in the built query string, but I would keep it to an explicit behaviour for `None`, and make sure you document it in the function definition.

Reply:
Aight. Updated it.